packages/loopring_v3/contracts/impl/libexchange/ExchangeBlocks.sol

```diff
@@ -114,192 +91,65 @@ library ExchangeBlocks
-                bytes32 startingHash = S.withdrawalChain[startIdx - 1].accumulatedHash;
-                bytes32 endingHash = S.withdrawalChain[startIdx + count - 1].accumulatedHash;
-                // Pad the block so it's full
-                for (uint i = count; i < _block.blockSize; i++) {
-                    endingHash = sha256(
-                        abi.encodePacked(
-                            endingHash,
-                            uint24(0),
-                            uint16(0),
-                            uint96(0)
-                        )
-                    );
-                }
-                bytes32 inputStartingHash = 0x0;
-                bytes32 inputEndingHash = 0x0;
-                assembly {
-                    inputStartingHash := mload(add(data, 100))
-                    inputEndingHash := mload(add(data, 132))
-                }
-                require(inputStartingHash == startingHash, "INVALID_STARTING_HASH");
-                require(inputEndingHash == endingHash, "INVALID_ENDING_HASH");
-                numWithdrawalRequestsCommitted += uint32(count);
-            }
-        } else if (_block.blockType == ExchangeData.BlockType.TRANSFER) {
-            validateConditionalTransfers(
-                S,
-                _block.data,
-                _block.auxiliaryData
-            );
-        } else if (_block.blockType != ExchangeData.BlockType.OFFCHAIN_WITHDRAWAL) {
-            revert("UNSUPPORTED_BLOCK_TYPE");
-        }
+        // Process conditional transactions
+        uint blockFeeETH = processConditionalTransactions(
+            S,
+            offset,
+            _block.data,
+            _block.auxiliaryData
+        );
 
-        // Process withdrawals
-        if (_block.blockType == ExchangeData.BlockType.ONCHAIN_WITHDRAWAL ||
-            _block.blockType == ExchangeData.BlockType.OFFCHAIN_WITHDRAWAL) {
-            uint start = 4 + 32 + 32;
-            if (_block.blockType == ExchangeData.BlockType.ONCHAIN_WITHDRAWAL) {
-                start += 32 + 32 + 4 + 4;
-            }
-            uint length = 8 * _block.blockSize;
-            bytes memory withdrawals = new bytes(0);
-            assembly {
-                withdrawals := add(data, start)
-                mstore(withdrawals, length)
-            }
-            S.distributeWithdrawals(withdrawals);
-        }
+        // Transfer the onchain block fee to the operator
+        feeRecipient.sendETHAndVerify(blockFeeETH, gasleft());
 
         // Emit an event
-        emit BlockSubmitted(S.numBlocksSubmitted, publicDataHash);
-
-        // Transfer the block fee to the fee recipient (for block types with fees paid onchain)
-        withdrawBlockFee(
-            S,
-            S.numBlocksSubmitted,
-            numDepositRequestsCommitted,
-            numWithdrawalRequestsCommitted,
-            feeRecipient
-        );
+        emit BlockSubmitted(S.blocks.length, publicDataHash, blockFeeETH);
```

Comment:
Yeah that makes sense. I would just add it. `blockFeeETH` is added because it's the only value that cannot easily be found anywhere else and is not easy to calculate. The other block info can also be found elsewhere easily.

Reply:
Yeah, doing 1 event/block is cheap so not really worried about that. Loads and stored from storage are the expensive operations (SLOAD = 800 gas, SSTORE = 20,000 gas for zero -> non-zero or 5,000 gas / non-zero to non-zero).